Release notes for 2020.1 (#1022)
pymor/pymor: docs/source/release_notes.rst

See `here <https://github.com/pymor/pymor/blob/master/AUTHORS.md>`__ for more
details.

Release highlights
^^^^^^^^^^^^^^^^^^

Model order reduction using artificial neural networks
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
With this release, we introduce a simple approach for non-intrusive model order
reduction to pyMOR that makes use of artificial neural networks
`[#1001] <https://github.com/pymor/pymor/pull/1001>`_. The method was first
described in [HU18]_ and only requires being able to compute solution snapshots of
the full-order |Model|. Thus, it can be applied to arbitrary (nonlinear) |Models| even when no
access to the model's |Operators| is possible.

Our implementation internally wraps `PyTorch <https://pytorch.org>`_ for the training and evaluation of
the neural networks. No knowledge of PyTorch or neural networks is required to apply the method.

New system analysis and linear algebra algorithms
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
So far, computing poles of an |LTIModel| was only supported by its
:meth:`~pymor.models.iosys.LTIModel.poles` method which uses a dense eigenvalue
solver and converts the operators to dense matrices.

> The new version adds :meth:`~pymor.algorithms.samdp.samdp`
> (`[#834] <https://github.com/pymor/pymor/pull/834>`_) and
> :meth:`~pymor.algorithms.eigs.eigs`
> (`[#880] <https://github.com/pymor/pymor/pull/880>`_) which can compute a subset

Author: ```suggestion
The new version adds :meth:`~pymor.algorithms.samdp.samdp`
`[#834] <https://github.com/pymor/pymor/pull/834>`_ and
:meth:`~pymor.algorithms.eigs.eigs`
`[#880] <https://github.com/pymor/pymor/pull/880>`_ which can compute a subset
```

> of the eigenvalues which are dominant in some sense.

Author: @pmli, this sounds like samdp and eigs do similar things. Is this intended?

Reviewer: Not really. I'll work on it a bit.

Reviewer: @sdrave What do you think about the new version?

Author: See below.

pyMOR 2020.1 contains contributions by Linus Balicki, Tim Keil, Hendrik Kleikamp
and Luca Mechelli. We are also happy to welcome Linus as a new main developer!
See `here <https://github.com/pymor/pymor/blob/master/AUTHORS.md>`__ for more
details.

Release highlights
^^^^^^^^^^^^^^^^^^

Model order reduction using artificial neural networks
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
With this release, we introduce a simple approach for non-intrusive model order
reduction to pyMOR that makes use of artificial neural networks
`[#1001] <https://github.com/pymor/pymor/pull/1001>`_. The method was first
described in [HU18]_ and only requires being able to compute solution snapshots of
the full-order |Model|. Thus, it can be applied to arbitrary (nonlinear) |Models| even when no
access to the model's |Operators| is possible.

Our implementation internally wraps `PyTorch <https://pytorch.org>`_ for the training and evaluation of
the neural networks. No knowledge of PyTorch or neural networks is required to apply the method.

New system analysis and linear algebra algorithms
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> So far, computing poles of an |LTIModel| was only supported by its
> :meth:`~pymor.models.iosys.LTIModel.poles` method which uses a dense eigenvalue
> solver and converts the operators to dense matrices.
> The new version adds the :meth:`~pymor.algorithms.eigs.eigs` method
> (`[#880] <https://github.com/pymor/pymor/pull/880>`_) for computing a few
> smallest/largest eigenvalues (similar to `scipy.sparse.linalg.eigs`, except
> working for generic operators).
> Furthermore, a new addition is the :meth:`~pymor.algorithms.samdp.samdp` method
> (`[#834] <https://github.com/pymor/pymor/pull/834>`_) which takes into account
> the input and output operators of an |LTIModel| to compute its dominant poles.

Author: The new version still confuses me. How about:

```suggestion
The new :meth:`~pymor.algorithms.eigs.eigs` method
(`[#880] <https://github.com/pymor/pymor/pull/880>`_) computes 
smallest/largest eigenvalues of an arbitary linear real |Operator|
using the implicitly restarted Arnoldi method [RL95]_. It can also
be used to solve generalized eigenvalue problems.

So far, computing poles of an |LTIModel| was only supported by its
:meth:`~pymor.models.iosys.LTIModel.poles` method, which uses a dense eigenvalue
solver and converts the operators to dense matrices.
The new :meth:`~pymor.algorithms.samdp.samdp` method
(`[#834] <https://github.com/pymor/pymor/pull/834>`_) implements the
subspace accelerated dominant pole (SAMDP) algorithm  [RM06]_,
which can be used to compute the dominant poles operators of an 
|LTIModel| with arbitrary (in particular sparse) system |Operators|
without relying on dense matrix operations.
```

Reviewer: Yes, that sounds better. I forgot about other uses of `eigs`...

  dict are the individual *parameters* of the corresponding |ParametricObject|.
  The items of a :class:`~pymor.parameters.base.Mu` dict are the associated
  *parameter values*.
- All parameters are now one-dimensional NumPy arrays.
- Instead of manually calling `build_parameter_type` in `__init__`, the |parameters|
  of a |ParametricObject| are now automatically inferred from the object's `__init__`
  arguments. The process can be customized using the new `parameters_own` and
  `parameters_internal` properties.
- `CubicParameterSpace` was renamed to |ParameterSpace| and is created using
  `parametric_object.parameters.space(ranges)`.

Further details can be found in `[#923] <https://github.com/pymor/pymor/pull/923>`_.
Also see `[#949] <https://github.com/pymor/pymor/pull/949>`_ and
`[#998] <https://github.com/pymor/pymor/pull/998>`_.

pyMOR tutorial collection
~~~~~~~~~~~~~~~~~~~~~~~~~
Hands-on tutorials provide a good opportunity for new users to get started with
a software library. In this release a variety of tutorials have been added which
introduce important pyMOR concepts and basic model order reduction methods. In
particular users can now learn about:

> - :ref:`tutorial01`.
> - :ref:`tutorial-rb`
> - :ref:`tutorial_bt`
> - :ref:`tutorial_mor_with_anns`
> - :ref:`tutorial_external_solver`

Author: ```suggestion
- :doc:`tutorial_builtin_discretizer`.
- :doc:`tutorial_basis_generation`
- :doc:`tutorial_bt`
- :doc:`tutorial_mor_with_anns`
- :doc:`tutorial_external_solver`
```

I believe :ref: does not work.

Author: @lbalicki, it seems that your changes in e2e60cd5006e2220f39d869073c97b651a1ff1c3 are somehow based on older versions of the tutorials ... We should probably revert that commit.

Author: Sorry that was wrong. It's just that this PR is a bit behind master.

pyMOR tutorial collection
~~~~~~~~~~~~~~~~~~~~~~~~~
Hands-on tutorials provide a good opportunity for new users to get started with
a software library. In this release a variety of tutorials have been added which
introduce important pyMOR concepts and basic model order reduction methods. In
particular users can now learn about:

- pyMOR's discretization toolkit (`[#866] <https://github.com/pymor/pymor/pull/866>`_).
- How to build a reduced basis with pyMOR (`[#1013] <https://github.com/pymor/pymor/pull/1013>`_).
- Reducing a heat equation using balanced truncation (`[#984] <https://github.com/pymor/pymor/pull/984>`_).
- Model order reduction with artificial neural networks (`[#1001] <https://github.com/pymor/pymor/pull/1001>`_).

> - Binding an external PDE solver to pyMOR (`[#1007] <https://github.com/pymor/pymor/pull/1007>`_).

Author: Maybe it would actually make more sense to link to the documentation here instead of the PRs?

Reviewer 2: Yes, that probably makes more sense
